pyfixmsg/reference.py

```diff
@@ -292,20 +312,24 @@ def add_group(self, count_tag, composition, insert_at=None):
             # Will sort by tag number after the sorted tags otherwise
 
 
-def _extract_sorting_key(definition, sorting_key=None):
+def _extract_sorting_key(definition, spec, sorting_key=None, index=0):
     """
     Retrieve the sorting key for an object.
     The sorting key is used to serialise tags in the order they appear in the spec.
     It is unclear whether that's required for the root of the message (aside from header and tail)
     but it is essential in repeating groups. This takes the safe approach of enforcing it at all
     levels.
     """
-    sorting_key = sorting_key or {35: -1, 10: int(10e9)}
+    if sorting_key is None:
+        sorting_key = {35: -1, 10: int(10e9)}
+        for index, item in enumerate(spec.header_tags):
+            sorting_key[item.tag] = index
+    start_index = index + 1
```

Comment:
Thanks for your PR, I have some comment for it. It seems good for FIX4.2 or FIX4.4, but Quick FIX 5.0 spec has no headers defined (they were defined in another transportation spec, we might support it later). So, I don't think it works for FIX5.0 or later version.
For example, my message is '8=FIX.4.2;9=97;35=B;215=1;216=1;146=2;55=EURUSD;55=EURGBP;10=000;', after parsing it and serializing the message, I found tag 8 and 9 are not in correct order.

Something could be considered:
1). In FixSpec::__init__(), self.header_tags = [...] **or [self.tags.by_tag(tag) for tag in HEADER_TAGS]**
HEADER_TAGS is pre-defined in this file, but it's for FIX4.2 so I'm afraid it needs a little change for FIX5.0, is there any idea to make is available for all versions?
2). HEADER_TAGS of FIX4.4/5.0 also has group, just enumerating spec.header_tags to get tags' index is not enough to handle repeating groups.

Reply:
Good point. So yes, updating ``HEADER_TAGS`` to fit FIXT1.0 solves your example, but doesn't support the Hops repeating group. I think that adding to the doc saying that particular group is not yet supported until such time that someone actively needs it would be fair; what do you think?

Reply:
Yes, no need to support repeating group in header for now. But this code should work for both FIX4.x and FIX5.x so I suggest the following code, then we could remove the if condition in **test_serialisation_header()**

 for index, item in enumerate(header_tags):
            if isinstance(item, FixTag):
                sorting_key[item.tag] = index
            else:
                sorting_key[item] = index

Reply:
You’re right, i’ll remove the if statement. It made me realise that header ordering without an explicit spec is broken, will fix that now.

Reply:
fixed in latest commit